onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu

```diff
@@ -746,6 +746,195 @@ void TorchMultinomialKernelLauncher(float* d_input,
                                                 d_presence_mask);
 }
 
+template <typename T>
+__global__ void PickGptPastStateKernelImpl(
+    T* __restrict__ past_state,
+    const T* __restrict__ present_state,
+    const int32_t* beam_indices,
+    int batch_beam,
+    int past_seq_length,
+    int max_seq_length,
+    int num_heads,
+    int head_size) {
+  int layer_id = blockIdx.z / 2;
+  int is_v = blockIdx.z % 2;
+  int batch_beam_id = blockIdx.y;
+  int seq_id = blockIdx.x;
+  int head_id = threadIdx.y;
+
+  int block_per_layer = 2 * batch_beam * num_heads * max_seq_length * head_size;
+  int block_per_head = past_seq_length * head_size;
+  int block_per_beam = num_heads * block_per_head;
+  int block_per_past_state = batch_beam * block_per_beam;
+
+  int input_offset = (layer_id * block_per_layer) + (is_v * block_per_past_state) +
+                     (beam_indices[batch_beam_id] * block_per_beam) +
+                     (head_id * block_per_head) + (seq_id * head_size);
+
+  int output_offset = (layer_id * block_per_layer) + (is_v * block_per_past_state) +
+                      (batch_beam_id * block_per_beam) +
+                      (head_id * block_per_head) + (seq_id * head_size);
+
+  for (int i = threadIdx.x; i < head_size; i += blockDim.x) {
```

Comment:
there is no need of a for loop

Reply:
That's true. For now, each thread only processes one element.  It is needed when we eventually support single threadblock's thread count < N * H. But in any case, I doubt this affects performance as the loop will bail out after one iteration.